File: zoomy/src/main/java/com/ablanco/zoomy/MotionUtils.java
package com.ablanco.zoomy;

import android.graphics.PointF;
import android.view.MotionEvent;

/**
 * Created by Álvaro Blanco Cabrero on 11/02/2017.
 * Zoomy.
 */
class MotionUtils {
    static float lastOffsetX = 0;
    static float lastOffsetY = 0;

    static void midPointOfEvent(PointF point, MotionEvent event, boolean breakFakeFinger){
        int divider = 1;
        float x;
        float y;

        if (event.getPointerCount() == 2) {
            divider = 2;

            if (breakFakeFinger) {
                lastOffsetX = lastOffsetX - event.getX(1);
                lastOffsetY = lastOffsetY - event.getY(1);
            }

            x = event.getX(0) + event.getX(1);
            y = event.getY(0) + event.getY(1);
        } else {
            x = event.getX(0);
            y = event.getY(0);
        }

        point.set(x / divider + lastOffsetX, y / divider + lastOffsetY);
    }
}

File: zoomy/src/main/java/com/ablanco/zoomy/MotionUtils.kt
package com.ablanco.zoomy

import android.graphics.PointF
import android.view.MotionEvent

/**
 * Created by Álvaro Blanco Cabrero on 11/02/2017.
 * Zoomy.
 */
internal object MotionUtils {
    private var savedPositionX = 0f
    private var savedPositionY = 0f

    private var newOffsetX = 0F
    private var newOffsetY = 0F
    fun midPointOfEvent(point: PointF, pointers: List<PointerInfo>, event: MotionEvent) {
        val x: Float
        val y: Float

        val actionMasked = event.actionMasked()
        when (actionMasked) {
//            MotionEvent.ACTION_POINTER_DOWN,
//            MotionEvent.ACTION_DOWN -> {
//                point[lastOffsetX] = lastOffsetY
//                return
//            }
//
//            MotionEvent.ACTION_POINTER_UP,
//            MotionEvent.ACTION_UP,
//            MotionEvent.ACTION_CANCEL -> {
//                return
//            }
        }

        when (pointers.size) {
            0 -> {
                point[savedPositionX] = savedPositionY
                return
            }
            1 -> {
                val lastPointer = pointers.last()
                if (actionMasked == MotionEvent.ACTION_MOVE && newOffsetX == newOffsetY) {
                    newOffsetX = lastPointer.rawX - savedPositionX
                    newOffsetY = lastPointer.rawY - savedPositionY
                }
                point[savedPositionX + lastPointer.rawX + newOffsetX] = savedPositionY + lastPointer.rawY + newOffsetY
                return
            }
            else -> {
                x = pointers[0].rawX + pointers[1].rawX
                y = pointers[0].rawY + pointers[1].rawY
            }
        }
        var pointX = x / 2
        var pointY = y / 2

        when (actionMasked) {
            MotionEvent.ACTION_POINTER_UP -> {
                savedPositionX = pointX
                savedPositionY = pointY
            }
            MotionEvent.ACTION_POINTER_DOWN -> {
//                pointX = savedXOffset
//                pointY = savedYOffset
            }
        }
//        Log.e("TEST", "XXX: $pointX, YYY: $pointY")
//        if (actionMasked == MotionEvent.ACTION_POINTER_UP) {
//            Log.d("TEST", "XXX: $pointX, YYY: $pointY")
//            Log.e("TEST", "XXX: $pointX, YYY: $pointY")
//        }
        point[pointX] = pointY
    }

    fun MotionEvent.actionMasked() = action and MotionEvent.ACTION_MASK

}